src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -268,52 +286,288 @@ internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSy
+                        variableAccess = new BoundLocal(node, localSymbol, null, declType);
+                        return;
+                    }
+                    else
+                    {
+                        // We should have the right binder in the chain for a script or interactive, so we use the field for the pattern.
+                        Debug.Assert(node.SyntaxTree.Options.Kind != SourceCodeKind.Regular);
+                        GlobalExpressionVariable expressionVariableField = LookupDeclaredField(singleVariableDesignation);
+                        DiagnosticBag tempDiagnostics = DiagnosticBag.GetInstance();
+                        expressionVariableField.SetType(declType, tempDiagnostics);
+                        tempDiagnostics.Free();
+                        BoundExpression receiver = SynthesizeReceiver(node, expressionVariableField, diagnostics);
+
+                        variableSymbol = expressionVariableField;
+                        variableAccess = new BoundFieldAccess(node, receiver, expressionVariableField, null, hasErrors);
+                        return;
+                    }
+                case DiscardDesignationSyntax _:
+                case null:
+                    variableSymbol = null;
+                    variableAccess = null;
+                    return;
                 default:
-                    throw ExceptionUtilities.UnexpectedValue(node.Designation.Kind());
+                    throw ExceptionUtilities.UnexpectedValue(designation.Kind());
             }
 
-            var designation = (SingleVariableDesignationSyntax)node.Designation;
-            var identifier = designation.Identifier;
-            SourceLocalSymbol localSymbol = this.LookupLocal(identifier);
+        }
 
-            if (localSymbol != (object)null)
+        TypeSymbol BindRecursivePatternType(TypeSyntax typeSyntax, TypeSymbol operandType, ref bool hasErrors, out BoundTypeExpression boundDeclType, DiagnosticBag diagnostics)
+        {
+            if (typeSyntax != null)
             {
-                if ((InConstructorInitializer || InFieldInitializer) && ContainingMemberOrLambda.ContainingSymbol.Kind == SymbolKind.NamedType)
+                boundDeclType = BindPatternType(typeSyntax, operandType, ref hasErrors, out var isVar, diagnostics);
+                if (isVar)
                 {
-                    Error(diagnostics, ErrorCode.ERR_ExpressionVariableInConstructorOrFieldInitializer, node);
+                    // The type `var` is not permitted in recursive patterns. If you want the type inferred, just omit it.
+                    if (!hasErrors)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_InferredRecursivePatternType, typeSyntax.Location);
+                    }
+
+                    boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt: null, inferredType: true, type: operandType.StrippedType(), hasErrors: hasErrors = true);
```

Comment:
>hasErrors = true [](start = 153, length = 16)

The code would be easier to follow and maintain if this assignment was a separate statement.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156514969](https://github.com/dotnet/roslyn/pull/23209#discussion_r156514969) [](ancestors = 156514969)